lib/identifiers/doi.rb

```diff
@@ -1,7 +1,31 @@
 module Identifiers
   class DOI
+    PATTERN = %r{
+      \b
+      10 # Directory indicator (always 10)
+      \.
+      (?:
+        # ISBN-A
+        97[89]\. # ISBN (GS1) Bookland prefix
+        \d{2,8}  # ISBN registration group element and publisher prefix
+        /        # Prefix/suffix divider
+        \d{1,7}  # ISBN title enumerator and check digit
+        |
+        # DOI
+        \d{4,9} # Registrant code
+        /       # Prefix/suffix divider
+        \S+     # DOI suffix
+      )
+    }x
+
     def self.extract(str)
-      str.scan(%r{\b10\.(?:97[89]\.\d{2,8}/\d{1,7}|\d{4,9}/\S+)\b}).map(&:downcase)
+      str
+        .scan(PATTERN)
+        .map { |doi|
+          next doi.downcase if doi !~ /\p{Punct}\z/ || doi =~ /\(.+\)\z/
```

Comment:
What happens if there's more than one contiguous punctuation character after the DOI? For example, if someone uses `...` (rather than the single glyph `…` or even gets over excited and uses `!!!`

In these cases, do we risk only trimming the final one?